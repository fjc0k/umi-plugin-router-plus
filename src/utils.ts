import {flat, includes, isEmpty, isObject, isString, memoize} from 'vtils'
import {INormalizedParam} from './types'
import {IRoute, utils} from 'umi'
import {Key, parse} from 'path-to-regexp'
import {PARAM_DESCRIPTOR_BOOLEAN_TYPE, PARAM_DESCRIPTOR_ENUM_SEPARATOR, PARAM_DESCRIPTOR_NUMBER_TYPE, PARAM_DESCRIPTOR_SEGMENT_SEPARATOR, PARAM_DESCRIPTOR_TYPES} from './consts'

const {upperFirst, camelCase} = utils.lodash

const parsePath = memoize(parse)

/**
 * 将路由表转换为只有一级。
 *
 * @param routes 路由表
 */
export function flatRoutes(routes: IRoute[]): IRoute[] {
  return flat(
    routes.map(route => [
      route,
      ...flatRoutes(route.routes || []),
    ]),
  )
}

/**
 * 遍历路由表。
 *
 * @param routes 路由表
 * @param cb 回调
 */
export function walkRoutes(routes: IRoute[], cb: (route: IRoute) => IRoute): IRoute[] {
  return routes.map(route => {
    route = cb(route)
    if (!isEmpty(route.routes)) {
      route.routes = walkRoutes(route.routes!, cb)
    }
    return route
  })
}

/**
 * 获取规范化的路由名称。
 *
 * - 获取 `pageName` 或 `name` 作为原始名称，没有则从 `path` 中解析；
 * - 对获取到的原始名称做 `PascalCase` 转换后作为基本名称，若其为空，则设为 Index；
 * - 判断是否有子路由，是则将后缀设为 `Layout`；
 * - 将基本名称和后缀相加作为规范化的路由名称返回。
 *
 * @param route 路由信息
 */
export function getNormalizedRouteName(route: IRoute): string {
  const originalName = route.pageName || route.name || (
    parsePath(route.path || '')
      .filter(isString)
      .join('')
      .replace(/\.html$/, '')
  )
  const baseName = upperFirst(camelCase(originalName)) || 'Index'
  const suffix = isEmpty(route.routes) ? '' : 'Layout'
  const routeName = `${baseName}${suffix}`
  return routeName
}

/**
 * 获取规范化的路由参数。
 *
 * @param route 路由信息
 */
export function getNormalizedRouteParams(route: IRoute): INormalizedParam[] {
  const tokens = parsePath(route.path || '')
  const params = (tokens.filter(isObject) as Key[]).map<INormalizedParam>(param => {
    return {
      ...param,
      ...getNormalizedRouteParamDescriptor(param as any),
      originalName: String(param.name),
    }
  })
  return params
}

/**
 * 获取规范化的路由参数描述。
 *
 * @param param 参数信息
 */
export const getNormalizedRouteParamDescriptor = memoize(
  (param: INormalizedParam): Pick<INormalizedParam, 'name' | 'jsType' | 'tsType'> => {
    const [name, ...descriptors] = String(param.name).split(PARAM_DESCRIPTOR_SEGMENT_SEPARATOR)

    let jsType: INormalizedParam['jsType']
    let tsType: INormalizedParam['tsType']

    // ==== 根据模式推测 ====
    // 数值
    if (param.pattern === '\\d+' || param.pattern === '[0-9]+') {
      jsType = Number
      tsType = 'number'
    }

    // 枚举数值
    else if (/^[0-9|]+$/.test(param.pattern)) {
      jsType = Number
      tsType = param.pattern.split('|')
        .map(value => Number(value))
        .join(' | ')
    }

    // 枚举
    else if (/^[A-Za-z0-9_|]+$/.test(param.pattern)) {
      jsType = String
      tsType = param.pattern.split('|')
        .map(value => JSON.stringify(value))
        .join(' | ')
    }

    // 默认
    else {
      jsType = String
      tsType = 'string'
    }

    // ==== 根据描述推测 ====
    if (descriptors.length > 0) {
      // == 类型 ==
      const type = descriptors.find(descriptor => includes(PARAM_DESCRIPTOR_TYPES as any, descriptor))
      // 布尔
      if (type === PARAM_DESCRIPTOR_BOOLEAN_TYPE) {
        jsType = Boolean
        tsType = 'boolean'
      }
      // 数字
      else if (type === PARAM_DESCRIPTOR_NUMBER_TYPE) {
        jsType = Number
        tsType = 'number'
      }
      // 字符串
      else {
        jsType = String
        tsType = 'string'
      }

      // == 枚举 ==
      const enums = (descriptors.find(descriptor => includes(descriptor, PARAM_DESCRIPTOR_ENUM_SEPARATOR)) || '').split(PARAM_DESCRIPTOR_ENUM_SEPARATOR)
      if (enums.length > 1) {
        // 数字
        if (jsType === Number) {
          tsType = enums.map(Number).join(' | ')
        }
        // 字符串
        else if (jsType === String) {
          tsType = enums.map(value => `'${value}'`).join(' | ')
        }
      }
    }

    // 重复
    jsType = param.repeat
      ? [jsType] as any
      : jsType
    tsType = param.repeat
      ? `${tsType} | Array<${tsType}>`
      : tsType

    return {name, jsType, tsType}
  },
  {
    serializer: param => `${param.name}###${param.repeat}###${param.pattern}`,
  },
)
