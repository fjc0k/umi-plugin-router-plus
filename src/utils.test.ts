import {flatRoutes, getNormalizedRouteName, getNormalizedRouteParams, walkRoutes} from './utils'
import {IRoute, utils} from 'umi'
import {omit} from 'vtils'

const routes: IRoute[] = [
  {
    path: '1',
    routes: [
      {
        path: '1.1',
        routes: [],
      },
      {
        path: '1.2',
        routes: [
          {
            path: '1.2.1',
          },
        ],
      },
    ],
  },
  {
    path: '2',
  },
  {
    path: '3',
    routes: [],
  },
  {
    path: '4',
    routes: [
      {
        path: '4.1',
        routes: [
          {
            path: '4.1.1',
            routes: [
              {
                path: '4.1.1.1',
                routes: [
                  {
                    path: '4.1.1.1.1',
                  },
                ],
              },
            ],
          },
        ],
      },
    ],
  },
]

describe('flatRoutes', () => {
  test('表现正常', () => {
    const _routes = utils.lodash.cloneDeep(routes)
    expect(flatRoutes(_routes)).toMatchSnapshot('一级路由表')
  })
})

describe('walkRoutes', () => {
  test('表现正常', () => {
    const _routes = utils.lodash.cloneDeep(routes)
    expect(walkRoutes(_routes, route => {
      route.path += '_ok'
      return route
    })).toMatchSnapshot('遍历路由表')
  })
})

describe('getNormalizedRouteName', () => {
  const route: IRoute = {
    pageName: 'isPageName',
    name: 'isName',
    path: 'isPath',
  }
  test('pageName 优先', () => {
    expect(
      getNormalizedRouteName(route),
    ).toMatch(
      utils.lodash.upperFirst(route.pageName),
    )
  })
  test('name 次之', () => {
    expect(
      getNormalizedRouteName(omit(route, ['pageName'])),
    ).toMatch(
      utils.lodash.upperFirst(route.name),
    )
  })
  test('path 最后', () => {
    expect(
      getNormalizedRouteName(omit(route, ['pageName', 'name'])),
    ).toMatch(
      utils.lodash.upperFirst(route.path),
    )
  })
  test('应去除 path 中的 .html 后缀', () => {
    expect(
      getNormalizedRouteName({path: 'isPath.html'}),
    ).toMatch('IsPath')
  })
  test('包含子路由时加上 Layout 后缀', () => {
    expect(
      getNormalizedRouteName({path: 'isPath.html', routes: []}),
    ).toMatch('IsPath')
    expect(
      getNormalizedRouteName({path: 'isPath.html', routes: [{}]}),
    ).toMatch('IsPathLayout')
  })
  test('基本名称为空时应将之设为 Index', () => {
    expect(
      getNormalizedRouteName({path: '/'}),
    ).toMatch('Index')
  })
})

describe('getNormalizedRouteParams', () => {
  test('表现正常', () => {
    expect(
      getNormalizedRouteParams({}),
    ).toMatchSnapshot('没有 path')
    expect(
      getNormalizedRouteParams({path: ''}),
    ).toMatchSnapshot('path 为空')
    expect(
      getNormalizedRouteParams({path: '/sss/ssk'}),
    ).toMatchSnapshot('没有参数')
    expect(
      getNormalizedRouteParams({path: '/:sss/ssk'}),
    ).toMatchSnapshot('一个没有指定类型的参数')
    expect(
      getNormalizedRouteParams({path: '/:sss?/ssk'}),
    ).toMatchSnapshot('一个没有指定类型的可选参数')
    expect(
      getNormalizedRouteParams({path: '/:sss+/ssk'}),
    ).toMatchSnapshot('一个没有指定类型的重复参数')
    expect(
      getNormalizedRouteParams({path: '/:sss__n/ssk'}),
    ).toMatchSnapshot('一个指定类型为数字的参数')
    expect(
      getNormalizedRouteParams({path: '/:sss(\\d+)/ssk'}),
    ).toMatchSnapshot('一个通过正则指定类型为数字的参数')
    expect(
      getNormalizedRouteParams({path: '/:sss__n?/ssk'}),
    ).toMatchSnapshot('一个指定类型为数字的可选参数')
    expect(
      getNormalizedRouteParams({path: '/:sss__n+/ssk'}),
    ).toMatchSnapshot('一个指定类型为数字的重复参数')
    expect(
      getNormalizedRouteParams({path: '/:sss__b/ssk'}),
    ).toMatchSnapshot('一个指定类型为布尔的参数')
    expect(
      getNormalizedRouteParams({path: '/:sss__s/ssk'}),
    ).toMatchSnapshot('一个指定类型为字符串的参数')
    expect(
      getNormalizedRouteParams({path: '/:sss__1_2_3/ssk'}),
    ).toMatchSnapshot('一个没有指定类型但指定了枚举的参数')
    expect(
      getNormalizedRouteParams({path: '/:sss__n__1_2_3/ssk'}),
    ).toMatchSnapshot('一个指定类型为数字且指定了枚举的参数')
    expect(
      getNormalizedRouteParams({path: '/post/:id(\\d+)/:pid([0-9]+)/type_:type(1|2)_:name__s/:gender__f_m/:is_18__b?/:prop(x|y|0)'}),
    ).toMatchSnapshot('一个复杂的综合案例')
  })
})
