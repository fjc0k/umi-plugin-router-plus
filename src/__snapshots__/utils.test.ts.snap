// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flatRoutes 表现正常: 一级路由表 1`] = `
Array [
  Object {
    "path": "1",
    "routes": Array [
      Object {
        "path": "1.1",
        "routes": Array [],
      },
      Object {
        "path": "1.2",
        "routes": Array [
          Object {
            "path": "1.2.1",
          },
        ],
      },
    ],
  },
  Object {
    "path": "1.1",
    "routes": Array [],
  },
  Object {
    "path": "1.2",
    "routes": Array [
      Object {
        "path": "1.2.1",
      },
    ],
  },
  Object {
    "path": "1.2.1",
  },
  Object {
    "path": "2",
  },
  Object {
    "path": "3",
    "routes": Array [],
  },
  Object {
    "path": "4",
    "routes": Array [
      Object {
        "path": "4.1",
        "routes": Array [
          Object {
            "path": "4.1.1",
            "routes": Array [
              Object {
                "path": "4.1.1.1",
                "routes": Array [
                  Object {
                    "path": "4.1.1.1.1",
                  },
                ],
              },
            ],
          },
        ],
      },
    ],
  },
  Object {
    "path": "4.1",
    "routes": Array [
      Object {
        "path": "4.1.1",
        "routes": Array [
          Object {
            "path": "4.1.1.1",
            "routes": Array [
              Object {
                "path": "4.1.1.1.1",
              },
            ],
          },
        ],
      },
    ],
  },
  Object {
    "path": "4.1.1",
    "routes": Array [
      Object {
        "path": "4.1.1.1",
        "routes": Array [
          Object {
            "path": "4.1.1.1.1",
          },
        ],
      },
    ],
  },
  Object {
    "path": "4.1.1.1",
    "routes": Array [
      Object {
        "path": "4.1.1.1.1",
      },
    ],
  },
  Object {
    "path": "4.1.1.1.1",
  },
]
`;

exports[`getNormalizedRouteParams 表现正常: path 为空 1`] = `Array []`;

exports[`getNormalizedRouteParams 表现正常: 一个复杂的综合案例 1`] = `
Array [
  Object {
    "asterisk": false,
    "delimiter": "/",
    "jsType": [Function Number],
    "name": "id",
    "optional": false,
    "originalName": "id",
    "partial": false,
    "pattern": "\\\\d+",
    "prefix": "/",
    "repeat": false,
    "tsType": "number",
  },
  Object {
    "asterisk": false,
    "delimiter": "/",
    "jsType": [Function Number],
    "name": "pid",
    "optional": false,
    "originalName": "pid",
    "partial": false,
    "pattern": "[0-9]+",
    "prefix": "/",
    "repeat": false,
    "tsType": "number",
  },
  Object {
    "asterisk": false,
    "delimiter": "/",
    "jsType": [Function Number],
    "name": "type",
    "optional": false,
    "originalName": "type",
    "partial": false,
    "pattern": "1|2",
    "prefix": "",
    "repeat": false,
    "tsType": "1 | 2",
  },
  Object {
    "asterisk": false,
    "delimiter": "/",
    "jsType": [Function String],
    "name": "name",
    "optional": false,
    "originalName": "name__s",
    "partial": false,
    "pattern": "[^\\\\/]+?",
    "prefix": "",
    "repeat": false,
    "tsType": "string",
  },
  Object {
    "asterisk": false,
    "delimiter": "/",
    "jsType": [Function String],
    "name": "gender",
    "optional": false,
    "originalName": "gender__f_m",
    "partial": false,
    "pattern": "[^\\\\/]+?",
    "prefix": "/",
    "repeat": false,
    "tsType": "'f' | 'm'",
  },
  Object {
    "asterisk": false,
    "delimiter": "/",
    "jsType": [Function Boolean],
    "name": "is_18",
    "optional": true,
    "originalName": "is_18__b",
    "partial": false,
    "pattern": "[^\\\\/]+?",
    "prefix": "/",
    "repeat": false,
    "tsType": "boolean",
  },
  Object {
    "asterisk": false,
    "delimiter": "/",
    "jsType": [Function String],
    "name": "prop",
    "optional": false,
    "originalName": "prop",
    "partial": false,
    "pattern": "x|y|0",
    "prefix": "/",
    "repeat": false,
    "tsType": "\\"x\\" | \\"y\\" | \\"0\\"",
  },
]
`;

exports[`getNormalizedRouteParams 表现正常: 一个指定类型为字符串的参数 1`] = `
Array [
  Object {
    "asterisk": false,
    "delimiter": "/",
    "jsType": [Function String],
    "name": "sss",
    "optional": false,
    "originalName": "sss__s",
    "partial": false,
    "pattern": "[^\\\\/]+?",
    "prefix": "/",
    "repeat": false,
    "tsType": "string",
  },
]
`;

exports[`getNormalizedRouteParams 表现正常: 一个指定类型为布尔的参数 1`] = `
Array [
  Object {
    "asterisk": false,
    "delimiter": "/",
    "jsType": [Function Boolean],
    "name": "sss",
    "optional": false,
    "originalName": "sss__b",
    "partial": false,
    "pattern": "[^\\\\/]+?",
    "prefix": "/",
    "repeat": false,
    "tsType": "boolean",
  },
]
`;

exports[`getNormalizedRouteParams 表现正常: 一个指定类型为数字且指定了枚举的参数 1`] = `
Array [
  Object {
    "asterisk": false,
    "delimiter": "/",
    "jsType": [Function Number],
    "name": "sss",
    "optional": false,
    "originalName": "sss__n__1_2_3",
    "partial": false,
    "pattern": "[^\\\\/]+?",
    "prefix": "/",
    "repeat": false,
    "tsType": "1 | 2 | 3",
  },
]
`;

exports[`getNormalizedRouteParams 表现正常: 一个指定类型为数字的参数 1`] = `
Array [
  Object {
    "asterisk": false,
    "delimiter": "/",
    "jsType": [Function Number],
    "name": "sss",
    "optional": false,
    "originalName": "sss__n",
    "partial": false,
    "pattern": "[^\\\\/]+?",
    "prefix": "/",
    "repeat": false,
    "tsType": "number",
  },
]
`;

exports[`getNormalizedRouteParams 表现正常: 一个指定类型为数字的可选参数 1`] = `
Array [
  Object {
    "asterisk": false,
    "delimiter": "/",
    "jsType": [Function Number],
    "name": "sss",
    "optional": true,
    "originalName": "sss__n",
    "partial": false,
    "pattern": "[^\\\\/]+?",
    "prefix": "/",
    "repeat": false,
    "tsType": "number",
  },
]
`;

exports[`getNormalizedRouteParams 表现正常: 一个指定类型为数字的重复参数 1`] = `
Array [
  Object {
    "asterisk": false,
    "delimiter": "/",
    "jsType": Array [
      [Function Number],
    ],
    "name": "sss",
    "optional": false,
    "originalName": "sss__n",
    "partial": false,
    "pattern": "[^\\\\/]+?",
    "prefix": "/",
    "repeat": true,
    "tsType": "number | Array<number>",
  },
]
`;

exports[`getNormalizedRouteParams 表现正常: 一个没有指定类型但指定了枚举的参数 1`] = `
Array [
  Object {
    "asterisk": false,
    "delimiter": "/",
    "jsType": [Function String],
    "name": "sss",
    "optional": false,
    "originalName": "sss__1_2_3",
    "partial": false,
    "pattern": "[^\\\\/]+?",
    "prefix": "/",
    "repeat": false,
    "tsType": "'1' | '2' | '3'",
  },
]
`;

exports[`getNormalizedRouteParams 表现正常: 一个没有指定类型的参数 1`] = `
Array [
  Object {
    "asterisk": false,
    "delimiter": "/",
    "jsType": [Function String],
    "name": "sss",
    "optional": false,
    "originalName": "sss",
    "partial": false,
    "pattern": "[^\\\\/]+?",
    "prefix": "/",
    "repeat": false,
    "tsType": "string",
  },
]
`;

exports[`getNormalizedRouteParams 表现正常: 一个没有指定类型的可选参数 1`] = `
Array [
  Object {
    "asterisk": false,
    "delimiter": "/",
    "jsType": [Function String],
    "name": "sss",
    "optional": true,
    "originalName": "sss",
    "partial": false,
    "pattern": "[^\\\\/]+?",
    "prefix": "/",
    "repeat": false,
    "tsType": "string",
  },
]
`;

exports[`getNormalizedRouteParams 表现正常: 一个没有指定类型的重复参数 1`] = `
Array [
  Object {
    "asterisk": false,
    "delimiter": "/",
    "jsType": Array [
      [Function String],
    ],
    "name": "sss",
    "optional": false,
    "originalName": "sss",
    "partial": false,
    "pattern": "[^\\\\/]+?",
    "prefix": "/",
    "repeat": true,
    "tsType": "string | Array<string>",
  },
]
`;

exports[`getNormalizedRouteParams 表现正常: 一个通过正则指定类型为数字的参数 1`] = `
Array [
  Object {
    "asterisk": false,
    "delimiter": "/",
    "jsType": [Function Number],
    "name": "sss",
    "optional": false,
    "originalName": "sss",
    "partial": false,
    "pattern": "\\\\d+",
    "prefix": "/",
    "repeat": false,
    "tsType": "number",
  },
]
`;

exports[`getNormalizedRouteParams 表现正常: 没有 path 1`] = `Array []`;

exports[`getNormalizedRouteParams 表现正常: 没有参数 1`] = `Array []`;

exports[`walkRoutes 表现正常: 遍历路由表 1`] = `
Array [
  Object {
    "path": "1_ok",
    "routes": Array [
      Object {
        "path": "1.1_ok",
        "routes": Array [],
      },
      Object {
        "path": "1.2_ok",
        "routes": Array [
          Object {
            "path": "1.2.1_ok",
          },
        ],
      },
    ],
  },
  Object {
    "path": "2_ok",
  },
  Object {
    "path": "3_ok",
    "routes": Array [],
  },
  Object {
    "path": "4_ok",
    "routes": Array [
      Object {
        "path": "4.1_ok",
        "routes": Array [
          Object {
            "path": "4.1.1_ok",
            "routes": Array [
              Object {
                "path": "4.1.1.1_ok",
                "routes": Array [
                  Object {
                    "path": "4.1.1.1.1_ok",
                  },
                ],
              },
            ],
          },
        ],
      },
    ],
  },
]
`;
